File: frontend/src/types/dimensionamento.ts
/**
 * Tipos TypeScript espelhando os schemas Pydantic da API
 * (backend/app/schemas/dimensionamento.py).
 *
 * Estes tipos sao apenas o contrato de I/O da API consumida pelo frontend.
 * Nenhuma regra de calculo e replicada aqui.
 */

// ---------------------------------------------------------------------
// Entrada (request)
// ---------------------------------------------------------------------

export type TipoItemConsumo = "AC" | "DC";

export interface ItemConsumoInput {
  descricao: string;
  tipo: TipoItemConsumo;
  potencia?: number | null;
  corrente?: number | null;
  uso_pct?: number | null;
  eficiencia_pct?: number | null;
}

export interface RetrofitInput {
  ah_chumbo: number;
  dod_chumbo: number;
  ef_chumbo: number;
  dod_lfp: number;
  ef_lfp: number;
}

export interface ControladorInput {
  v_min: number;
  v_max: number;
  i_cont: number;
  i_pico: number;
}

export interface DimensionamentoRequest {
  tensao: number;
  autonomia: number;
  fator: number;
  itens_consumo: ItemConsumoInput[];
  modo_selecao: string;
  retrofit?: RetrofitInput | null;
  controlador?: ControladorInput | null;
}

// ---------------------------------------------------------------------
// Saida (response)
// ---------------------------------------------------------------------

export interface ResumoDimensionamento {
  potencia_total: number;
  i_max: number;
  i_media: number;
  ah_por_consumo: number;
  ah_necessario: number;
  kwh_necessario: number;
  serie: number;
  v_nom: number;
  v_max: number;
  v_min: number;
}

export interface RetrofitResultado {
  ah_real_chumbo: number;
  ah_lfp: number;
}

export interface AlertaControlador {
  nivel: "warning" | "error" | string;
  mensagem: string;
}

export interface OpcaoCelula {
  fabricante: string;
  ah: number;
  c_cont: number;
  c_pico: number;
  cont_datasheet: number;
  pico_datasheet: number;
  cont: number;
  pico: number;
  cont_recomendado: number;
  pico_recomendado: number;
  peso: number;
  ciclos: number;
  condicao_ciclos: string;
  comprimento_mm: number;
  largura_mm: number;
  altura_mm: number;
  serie: number;
  paralelo: number;
  total_celulas: number;
  capacidade_pack: number;
  energia_pack: number;
  cont_pack: number;
  pico_pack: number;
  cont_datasheet_pack: number;
  pico_datasheet_pack: number;
  peso_pack: number;
  autonomia: number;
  c_rate_cont: number;
  c_rate_pico: number;
  c_rate_uso: number;
}

export interface ComparativoLinha {
  celula: string;
  configuracao: string;
  ah_final: number;
  kwh: number;
  c_rate_continuo: number;
  c_rate_pico: number;
  continua_datasheet_a: number;
  continua_fullenergy_a: number;
  pico_datasheet_a: number;
  pico_fullenergy_a: number;
  c_rate_utilizado: number;
  peso_celulas_kg: number;
  autonomia_h: number | null;
}

export interface DimensionamentoResponse {
  resumo: ResumoDimensionamento;
  retrofit?: RetrofitResultado | null;
  opcoes: OpcaoCelula[];
  comparativo: ComparativoLinha[];
  celula_selecionada?: OpcaoCelula | null;
  alertas_controlador: AlertaControlador[];
}

// ---------------------------------------------------------------------
// Tipos auxiliares de formulario (uso exclusivo do frontend)
// ---------------------------------------------------------------------

/** Tipo de projeto selecionado na Secao 1 (Dados do Projeto). */
export type TipoProjeto = "novo" | "retrofit";

/** Modo de selecao de celula na Secao 5. */
export type ModoSelecaoUI = "automatica" | "manual";

/**
 * Linha editavel da tabela de Consumo (Secao 2). Mantem um `id` local para
 * controle de chaves React; os demais campos correspondem a
 * `ItemConsumoInput`.
 */
export interface ItemConsumoFormulario {
  id: string;
  descricao: string;
  tipo: TipoItemConsumo;
  potencia: number;
  corrente: number;
  uso_pct: number;
  eficiencia_pct: number;
}

File: backend/app/schemas/dimensionamento.py
"""
Schemas (Pydantic) da API de dimensionamento.

Define os formatos de entrada e saida do endpoint POST /dimensionar.
Estes schemas sao apenas contratos de I/O da API -- nao contem nenhuma
regra de calculo, que permanece exclusivamente em `app.calculo`.
"""

from typing import List, Literal, Optional

from pydantic import BaseModel, Field


class ItemConsumoInput(BaseModel):
    """Uma linha de consumo (motor ou componente auxiliar).

    Campos opcionais nao informados (None) sao tratados pela camada de
    calculo (`app.calculo.consumo`) com os mesmos defaults do app atual:
    potencia=0, corrente=0, uso_pct=100, eficiencia_pct=90.
    """

    descricao: str = ""
    tipo: Literal["AC", "DC"] = "DC"
    potencia: Optional[float] = None
    corrente: Optional[float] = None
    uso_pct: Optional[float] = None
    eficiencia_pct: Optional[float] = None


class RetrofitInput(BaseModel):
    """Dados do projeto de retrofit (chumbo -> LiFePO4).

    Valores padrao iguais aos da tela de retrofit do app atual:
    ah_chumbo=220, dod_chumbo=80, ef_chumbo=70, dod_lfp=95, ef_lfp=95.
    """

    ah_chumbo: float = 220.0
    dod_chumbo: float = 80.0
    ef_chumbo: float = 70.0
    dod_lfp: float = 95.0
    ef_lfp: float = 95.0


class ControladorInput(BaseModel):
    """Limites do controlador, usados em `validar_controlador`.

    Valor 0 (default) significa "nao informado" e a verificacao
    correspondente e ignorada -- igual ao app atual.
    """

    v_min: float = 0
    v_max: float = 0
    i_cont: float = 0
    i_pico: float = 0


class DimensionamentoRequest(BaseModel):
    """Corpo da requisicao POST /dimensionar."""

    tensao: float = Field(..., description="Tensao nominal do sistema (V)")
    autonomia: float = Field(0, description="Autonomia desejada (h)")
    fator: float = Field(0, description="Fator de utilizacao media (%)")
    itens_consumo: List[ItemConsumoInput] = Field(
        default_factory=list,
        description="Lista de motores e componentes auxiliares (projeto novo)",
    )
    modo_selecao: str = Field(
        "Automática",
        description=(
            'Modo de selecao da celula: "Automática" (selecao automatica) ou '
            'o identificador "{fabricante} {ah}Ah" para selecao manual'
        ),
    )
    retrofit: Optional[RetrofitInput] = Field(
        None, description="Preencher apenas em projetos de retrofit"
    )
    controlador: Optional[ControladorInput] = Field(
        None, description="Limites do controlador, para validacao (opcional)"
    )


class ResumoDimensionamento(BaseModel):
    potencia_total: float
    i_max: float
    i_media: float
    ah_por_consumo: float
    ah_necessario: float
    kwh_necessario: float
    serie: int
    v_nom: float
    v_max: float
    v_min: float


class RetrofitResultado(BaseModel):
    ah_real_chumbo: float
    ah_lfp: float


class AlertaControlador(BaseModel):
    nivel: str
    mensagem: str


class OpcaoCelula(BaseModel):
    fabricante: str
    ah: float
    c_cont: float
    c_pico: float
    cont_datasheet: float
    pico_datasheet: float
    cont: float
    pico: float
    cont_recomendado: float
    pico_recomendado: float
    peso: float
    ciclos: int
    condicao_ciclos: str
    comprimento_mm: int
    largura_mm: int
    altura_mm: int
    serie: int
    paralelo: int
    total_celulas: int
    capacidade_pack: float
    energia_pack: float
    cont_pack: float
    pico_pack: float
    cont_datasheet_pack: float
    pico_datasheet_pack: float
    peso_pack: float
    autonomia: float
    c_rate_cont: float
    c_rate_pico: float
    c_rate_uso: float


class ComparativoLinha(BaseModel):
    celula: str
    configuracao: str
    ah_final: float
    kwh: float
    c_rate_continuo: float
    c_rate_pico: float
    continua_datasheet_a: float
    continua_fullenergy_a: float
    pico_datasheet_a: float
    pico_fullenergy_a: float
    c_rate_utilizado: float
    peso_celulas_kg: float
    autonomia_h: Optional[float] = None


class DimensionamentoResponse(BaseModel):
    resumo: ResumoDimensionamento
    retrofit: Optional[RetrofitResultado] = None
    opcoes: List[OpcaoCelula]
    comparativo: List[ComparativoLinha]
    celula_selecionada: Optional[OpcaoCelula] = None
    alertas_controlador: List[AlertaControlador] = []
